Implement SMS commands

Abyss777 opened the pull request:
Unfortunately a lot of changes at one pull-request...

Key changes:

- Added map `protocol name` -> `protocol` for all protocols in `ServerManager`, we need protocol object access at any time. `deviceId` -> `lastPosition.getProtocol()` -> `protocol`
- Extended `BaseProtocol` to set list of supported sms commands and send sms command function and the main is store sms encoder
- `BaseProtocolSmsEncoder` extends `StringProtocolEncoder` there are few benefits:

  - We can use `initDevicePassword` and `formatCommand` functions for sms
  -  We can combine gprs and sms commands in one encoder in case they are the same or similar

- Implemented `sendCommand` and `getSupportedCommands` functions in `DeviceManager`
- Faced with one problem, my device (and I think most of other) understand commands only in ASCII or GSM encoding, but I think most of users want get notifications in their language (unicode UCS2). I had to introduce separate charset config parameters for notifications and for commands in `SmppClient`
- There are two example implementations:
  - Wondex has the same SMS format and encoder combined in one
  - Granit has different set of SMS commnds and encoders are split 

Other comments in the code...

Abyss777 commented on src/org/traccar/api/resource/CommandTypeResource.java:
```diff
@@ -34,9 +34,9 @@
 
     @GET
-    public Collection<CommandType> get(@QueryParam("deviceId") long deviceId) throws SQLException {
+    public Collection<CommandType> get(@QueryParam("deviceId") long deviceId, @QueryParam("sms") boolean sms) {
```
The same request for sms and gprs commands

Abyss777 commented on src/org/traccar/database/ActiveDevice.java:
```diff
@@ -47,16 +43,6 @@ public long getDeviceId() {
         return deviceId;
     }
 
-    public Collection<CommandType> getCommandTypes() {
```
It is unified in `DeviceManager.getSupportedCommands`

Abyss777 commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -420,4 +427,46 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
+            if (activeDevice != null) {
+                activeDevice.sendCommand(command);
+            } else {
+                throw new RuntimeException("Device is not online");
```
Add human-readable exception in case send gprs command to offline device

Abyss777 commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -420,4 +427,46 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
+                result.add(new CommandType(commandKey));
+            }
+        } else {
+            result.add(new CommandType(Command.TYPE_CUSTOM));
```
If device was never online, we still will able send custom sms command (for initial configuration for example)

Abyss777 commented on src/org/traccar/protocol/WondexProtocolEncoder.java:
```diff
@@ -40,4 +40,9 @@ protected Object encodeCommand(Command command) {
 
+    @Override
+    protected String encodeSmsCommand(Command command) {
+        return (String) encodeCommand(command);
```
Both formats are the same

tananaev replied:
I like that it can be reused.

Abyss777 commented on src/org/traccar/smpp/ClientSmppSessionHandler.java:
```diff
@@ -44,14 +44,14 @@ public PduResponse firePduRequestReceived(PduRequest request) {
                             + request.getOptionalParameter(SmppConstants.TAG_MSG_STATE).getValueAsByte());
                 } else {
-                    Log.debug("Message Received: "
+                    Log.debug("SMS Message Received: "
```
Small visual improvement, to understand what is message received.

Abyss777 commented on src/org/traccar/smpp/ClientSmppSessionHandler.java:
```diff
@@ -44,14 +44,14 @@ public PduResponse firePduRequestReceived(PduRequest request) {
+                    Log.debug("SMS Message Received: "
                             + CharsetUtil.decode(((DeliverSm) request).getShortMessage(),
-                            smppClient.mapDataCodingToCharset(((DeliverSm) request).getDataCoding()))
+                            smppClient.mapDataCodingToCharset(((DeliverSm) request).getDataCoding())).trim()
```
My wondex device add new line at the end of message, log looks not very nice.

Abyss777 commented on src/org/traccar/smpp/SmppClient.java:
```diff
@@ -184,31 +193,41 @@ private void destroySession() {
 
-    public synchronized void sendMessageSync(String destAddress, String message) throws RecoverablePduException,
-            UnrecoverablePduException, SmppTimeoutException, SmppChannelException, InterruptedException {
+    public synchronized void sendMessageSync(String destAddress, String message, boolean command)
```
`command` flag to choose charset

Abyss777 commented on src/org/traccar/smpp/SmppClient.java:
```diff
@@ -184,31 +193,41 @@ private void destroySession() {
+            if (submitResponce.getCommandStatus() == SmppConstants.STATUS_OK) {
+                Log.debug("SMS submited, msg_id: " + submitResponce.getMessageId());
+            } else {
+                throw new IllegalStateException(submitResponce.getResultMessage());
```
In some cases communication with smpp server can be successful, but really something wrong. For example destination address. We need to log this.

tananaev commented on src/org/traccar/api/resource/CommandResource.java:
```diff
@@ -32,10 +37,11 @@
 
     @POST
-    public Response add(Command entity) {
+    public Response add(Command entity) throws RecoverablePduException, UnrecoverablePduException,
```
Maybe we should just declare generic exception here.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -420,4 +427,46 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
         }
     }
+
+    public void sendCommand(Command command) throws RecoverablePduException, UnrecoverablePduException,
```
Same. I think it's better to throw generic exception.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -420,4 +427,46 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
+        if (lastPosition != null) {
+            Collection<String> commands;
+            if (sms) {
+                commands = Context.getServerManager().getProtocol(lastPosition.getProtocol()).getSupportedSmsCommands();
```
I think the code will be more readable if you get protocol once and then use ternary operator to get corresponding commands list.

tananaev commented on src/org/traccar/smpp/SmppClient.java:
```diff
@@ -184,31 +193,41 @@ private void destroySession() {
-
             SubmitSm submit = new SubmitSm();
+            byte[] textBytes;
+            if (command) {
```
I think ternary operator would be easier here as well.

tananaev commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -420,4 +427,46 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
+
+    public void sendCommand(Command command) throws RecoverablePduException, UnrecoverablePduException,
+            SmppTimeoutException, SmppChannelException, InterruptedException {
+        if (command.getSms()) {
```
Should we automatically try SMS command if device is not online? Possibly a configuration option.

tananaev commented on src/org/traccar/BaseProtocolSmsEncoder.java:
```diff
@@ -0,0 +1,30 @@
+
+public abstract class BaseProtocolSmsEncoder extends StringProtocolEncoder {
+
+    protected abstract String encodeSmsCommand(Command command);
```
Do we need a separate method? I think we never share encoders anyway, right?

Abyss777 commented on src/org/traccar/database/DeviceManager.java:
```diff
@@ -420,4 +426,48 @@ public void resetTotalDistance(DeviceTotalDistance deviceTotalDistance) throws S
+            } else {
+                if (fallbackToSms) {
+                    command.setSms(true);
+                    sendCommand(command);
```
Don't you mind the recursion?

Abyss777 commented:
`BaseProtocolSmsEncoder` is not needed now.
Fixed all other comments.
Added `command.fallbackToSms` config to allow fallback to sms command if device is not online

tananaev commented on src/org/traccar/smpp/SmppClient.java:
```diff
@@ -184,31 +193,36 @@ private void destroySession() {
             SubmitSmResp submitResponce = getSession().submit(submit, submitTimeout);
-            Log.debug("SMS submited, msg_id: " + submitResponce.getMessageId());
+            if (submitResponce.getCommandStatus() == SmppConstants.STATUS_OK) {
+                Log.debug("SMS submited, msg_id: " + submitResponce.getMessageId());
```
There is a typo here. Should be "submitted". Also, I think `msg_id` should be removed or converted to something more readable, like "id" or "message id".

tananaev commented on src/org/traccar/BaseProtocol.java:
```diff
@@ -67,4 +86,25 @@ public void sendCommand(ActiveDevice activeDevice, Command command) {
+                        "Command " + command.getType() + " is not supported in protocol " + getName());
+            }
+        } else {
+            throw new RuntimeException("Smpp client is not enabled");
```
Change "Smpp" to "SMPP" here. Camel-case naming convention applies to Java code, but log messages should follow English language rules. SMPP is an abbreviation, so should be all upper case, same as SMS.

tananaev commented on src/org/traccar/smpp/SmppClient.java:
```diff
@@ -184,31 +193,36 @@ private void destroySession() {
+                throw new IllegalStateException(submitResponce.getResultMessage());
+            }
         } else {
             throw new SmppChannelException("Smpp session is not connected");
```
Change "Smpp" to "SMPP" here. In the log message, not Java class name.

tananaev commented on src/org/traccar/Protocol.java:
```diff
@@ -16,4 +21,9 @@
 
+    Collection<String> getSupportedSmsCommands();
+
+    void sendSmsCommand(String phone, Command command) throws RecoverablePduException, UnrecoverablePduException,
```
I think it's better to declare generic `Exception` here as well. Otherwise too many things to catch. Maybe we should create an issue in cloudhopper repository to suggest a base cloudhopper exception for all of their exceptions.

Abyss777 commented:
OK. Sorry.
Fixed log messages and combined exceptions in `Protocol`

Abyss777 commented:
Am I missing the comment? Do not see it

tananaev commented on src/org/traccar/BaseProtocol.java:
```diff
@@ -67,4 +86,25 @@ public void sendCommand(ActiveDevice activeDevice, Command command) {
+    }
+
+    @Override
+    public void sendSmsCommand(String phone, Command command) throws RecoverablePduException, UnrecoverablePduException,
```
The comment was here. Please use generic Exception here as well.

Abyss777 commented:
It's done.

tananaev commented:
Merged, thanks.

Abyss777 commented:
What do you think about renaming some functions in `Protocol` and `BaseProtocol` ?
For example call usual commands `GprsCommands` or `ChannelCommands` 
`Protocol.sendCommand()` -> `Protocol.sendChannelCommand()`
`BaseProtocol.setSupportedCommands` -> `BaseProtocol.setSupportedChannelCommands`

If protocol has the same set of commands for GPRS and SMS then we can have `setSupportedCommands` to set both types.

The only thing is that renaming will touch all protocols.

tananaev commented:
Maybe we should use terms "data" and "text" instead of "GPRS" and "SMS" correspondingly?

Abyss777 commented:
Good idea, we won't be locked on transport

Abyss777 commented:
Should we also rename `Command` model filed `boolean sms` to `boolean text`? It might be confusing for API users.

tananaev commented:
We can rename the field to something like `textChannel` in the `Command`.

Abyss777 commented:
Should we rename all protocol encoders from `*ProtocolEncoder` to `*ProtocolDataEncoder` ?
And use `*ProtocolTextEncoder` for SMS encoders

tananaev commented:
I would say no... there would be too many changes. Maybe at some point in future when we have many text encoders.
